invenio_sip2/records/record.py

```diff
@@ -0,0 +1,217 @@
+        if data:
+            return cls(data)
+
+    @classmethod
+    def get_all_records(cls):
+        """Get all records."""
+        return [cls(obj) for obj in datastore.all(cls.record_type)]
+
+    def dumps(self, **kwargs):
+        """Return pure Python dictionary with record metadata."""
+        return deepcopy(dict(self))
+
+
+class Server(Sip2RecordMetadata):
+    """class forSIP2 server."""
+
+    record_type = 'server'
+
+    @property
+    def number_of_clients(self):
+        """Get number of clients."""
+        return len(self.get_clients())
+
+    @property
+    def is_running(self):
+        """Check if server is running."""
+        return self.get('status') == 'running'
+
+    def delete(self):
+        """Delete server and all attached clients."""
+        for data in self.get_clients():
+            Client(data).delete()
+        super().delete()
+
+    def get_clients(self):
+        """Return clients."""
+        filter_query = 'server:{server_id}'.format(
+            server_id=self.id
+        )
+        return self.search(
+            index_type=Client.record_type,
+            filter_query=filter_query
+        )
+
+    @classmethod
+    def create(cls, data, id_=None, **kwargs):
+        """Create record.
+
+        :param data: Dict with metadata.
+        :param id_: Specify a UUID to use for the new record.
+        """
+        # check if server already exist in datastore
+        server = cls.find_server(**data)
+        if server:
+            # check if server running
+            if server.is_running:
+                raise ServerAlreadyRunning(
+                    'ServerAlreadyRunning {id}'.format(id=server.id)
+                )
+            return server
+
+        return super().create(data, id_=id_)
+
+    @classmethod
+    def find_server(cls, **kwargs):
+        """Find server depending kwargs."""
+        for server in datastore.all(cls.record_type):
+            if kwargs.items() <= server.items():
+                # true only if `first` is a subset of `second`
+                return cls(server)
+
+
+class Client(Sip2RecordMetadata):
+    """class for selfcheck client."""
+
+    record_type = 'client'
+
+    def get_key(self):
+        """Get generated key for Client object."""
+        return '{record_type}:{id}_server:{server_id}'.format(
+            record_type=self.record_type,
+            id=self.id,
+            server_id=self.server_id
+        )
+
+    @property
+    def server_id(self):
+        """Get server identifier."""
+        return self.get('server').get('id')
+
+    def get_server(self):
+        """Get server object."""
+        return Server.get_record_by_id(self.server_id)
+
+    @property
+    def remote_app(self):
+        """Shortcut for remote app."""
+        return self.get_server().get('remote_app')
+
+    @property
+    def is_authenticated(self):
+        """Shortcut to check if the selfcheck client is authenticated."""
+        return self.get('authenticated')
```

Comment:
is `authenticated` is a required field ? if not, add a default value for `get` call

Reply:
done